In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, accuracy_score
import time

In [ ]:
data = pd.read_csv("heart_attack.csv")
data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

#### **Описание датасета**

Этот набор данных даёт представление о классификации сердечных приступов.
Этот набор данных идеально подходит для создания моделей классификации.

Характеристики:
- **age**: Возраст пациента
- **sex**: Пол пациента
- **cp**: Тип боли в груди
    - 0 - типичная стенокардия;
    - 1 - атипичная стенокардия;
    - 2 - боль, не связанная со стенокардией;
    - 3 - бессимптомное течение, артериальное давление в покое.
- **trtbps**: Артериальное давление в покое (в мм рт. ст.)
- **chol**: Холестерин в мг/дл, полученный с помощью датчика ИМТ
- **fbs**: (уровень сахара в крови натощак > 120 мг/дл) (1 = да; 0 = нет)
- **restecg**: Результаты электрокардиограммы в состоянии покоя
    - 0 - в норме,
    - 1 - наличие аномалии зубца ST-T (инверсия зубца T и/или подъем или депрессия сегмента ST > 0,05 мВ),
    - 2 - наличие вероятной или определенной гипертрофии левого желудочка по критериям Эстеса.
- **thalachh**: Максимальная достигнутая частота сердечных сокращений
- **exng**: Ангина, вызванная физической нагрузкой (1 = да; 0 = нет)
- **oldpeak**: Предыдущий пик
- slp
- **caa**: количество крупных сосудов (0-3)
- **thall** - коэффициент
- **output**: целевая переменная

#### Предобработка данных

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
data.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng     oldpeak         slp         caa  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

            thall      output  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [ ]:
# количество пропущенных значений в каждом столбце
data.isna().sum().sort_values(ascending=False)

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

In [ ]:
data = data.drop_duplicates()

X = data.drop(columns=['output'])
y = data['output']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### Bagging (Случайный лес)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

start_time = time.time()
rf = RandomForestClassifier(
    n_estimators=100,
    max_depth=10,
    min_samples_leaf=2,
    random_state=42,
    n_jobs=-1
)
rf.fit(X_train_scaled, y_train)
rf_time = time.time() - start_time

y_pred_rf = rf.predict(X_test_scaled)
rf_acc = accuracy_score(y_test, y_pred_rf)

print(f"Время обучения: {rf_time:.2f} сек")
print(f"Accuracy: {rf_acc:.4f}")
print(classification_report(y_test, y_pred_rf))

=== Bagging (Random Forest) ===
Время обучения: 0.36 сек
Accuracy: 0.8033
              precision    recall  f1-score   support

           0       0.86      0.68      0.76        28
           1       0.77      0.91      0.83        33

    accuracy                           0.80        61
   macro avg       0.82      0.79      0.80        61
weighted avg       0.81      0.80      0.80        61



### Boosting (Gradient Boosting)

In [ ]:
from xgboost import XGBClassifier

start_time = time.time()
xgb = XGBClassifier(
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1,
    random_state=42,
    use_label_encoder=False,
    eval_metric='logloss'
)
xgb.fit(X_train_scaled, y_train)
xgb_time = time.time() - start_time

y_pred_xgb = xgb.predict(X_test_scaled)
xgb_acc = accuracy_score(y_test, y_pred_xgb)

print(f"Время обучения: {xgb_time:.2f} сек")
print(f"Accuracy: {xgb_acc:.4f}")
print(classification_report(y_test, y_pred_xgb))

/usr/local/lib/python3.12/dist-packages/xgboost/training.py:199: UserWarning: [10:59:41] WARNING: /workspace/src/learner.cc:790: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Время обучения: 0.88 сек
Accuracy: 0.7705
              precision    recall  f1-score   support

           0       0.82      0.64      0.72        28
           1       0.74      0.88      0.81        33

    accuracy                           0.77        61
   macro avg       0.78      0.76      0.76        61
weighted avg       0.78      0.77      0.77        61



### Сравнение методов

In [ ]:
results = pd.DataFrame({
    "Метод": ["Bagging (Random Forest)", "Boosting (XGBoost)"],
    "Время (сек)": [rf_time, xgb_time],
    "Accuracy": [rf_acc, xgb_acc]
})

print(results.to_string(index=False))

                  Метод  Время (сек)  Accuracy
Bagging (Random Forest)     0.362935  0.803279
     Boosting (XGBoost)     0.881645  0.770492
